components/eamxx/src/diagnostics/entrainment_budget.cpp

```diff
@@ -0,0 +1,300 @@
+              if(pblinvalg == 1) {
+                // pblinvalg = 1 ---> based solely on
+                // d(T_mid)/d(p_mid); finding the min (because
+                // d(p_mid) < 0), so keeping signs
+                tm_grad(k) = (tm_icol(k - 1) - tm_icol(k)) / pm_diff;
+              } else if(pblinvalg == 2) {
+                // pblinvalg = 2 ---> based solely on
+                // d(theta_l)/d(p_mid); finding the min
+                // (because d(p_mid) < 0), so keeping signs
+                tm_grad(k) = (tl_icol(k - 1) - tl_icol(k)) / pm_diff;
+              } else if(pblinvalg == 3) {
+                // pblinvalg = 3 ---> based solely on
+                // d(q_t)/d(p_mid); finding the max (because
+                // d(p_mid) < 0), so reversing signs
+                tm_grad(k) = -(qt_icol(k - 1) - qt_icol(k)) / pm_diff;
+              }
+            });
+        team.team_barrier();
+
+        // Find minimum gradient, because d(p_mid) < 0 in definition above
+        // Starting from the surface, and ensuring p_mid > 70000.0 Pa,
+        // to avoid resolving to some odd place higher up in the atmosphere.
+        using minloc_t       = Kokkos::MinLoc<Real, int>;
+        using minloc_value_t = typename minloc_t::value_type;
+        minloc_value_t minloc;
+        Kokkos::parallel_reduce(
+            Kokkos::TeamVectorRange(team, 1, num_levs),
+            [&](const int &k, minloc_value_t &result) {
+              if(tm_grad(k) < result.val && pm_icol(k) > 70000.0) {
+                result.val = tm_grad(k);
+                result.loc = k;
+              }
+            },
+            minloc_t(minloc));
+        team.team_barrier();
+        opt_tm_grad_lev = minloc.loc;
+
+        // Save some outputs just above the "mixed" PBL
+        o_pm_hplus(icol) = pm_icol(opt_tm_grad_lev - 1);
+        o_tl_hplus(icol) = tl_icol(opt_tm_grad_lev - 1);
+        o_qt_hplus(icol) = qt_icol(opt_tm_grad_lev - 1);
+
+        // Save the dF term (F(h) - F(0))
+        o_df_inpbl(icol) = (sd_icol(opt_tm_grad_lev - 1) - sd_icol(0)) -
+                           (su_icol(opt_tm_grad_lev - 1) - su_icol(0)) +
+                           (ld_icol(opt_tm_grad_lev - 1) - ld_icol(0)) -
+                           (lu_icol(opt_tm_grad_lev - 1) - lu_icol(0));
+
+        // Now only need to compute below from opt_tm_grad_lev to num_levs
+        Kokkos::parallel_for(
+            Kokkos::TeamVectorRange(team, opt_tm_grad_lev, num_levs),
+            [&](int k) {
+              // The time-based tendencies
+              auto tl_tend = (tl_icol(k) - prev_tliq_icol(k)) / dt;
+              auto qt_tend = (qt_icol(k) - prev_qtot_icol(k)) / dt;
+              // the integrated (averaged) quantities
+              o_tl_caret(icol) += tl_icol(k) * pd_icol(k) / g;
+              o_tl_ttend(icol) += tl_tend * pd_icol(k) / g;
+              o_qt_caret(icol) += qt_icol(k) * pd_icol(k) / g;
+              o_qt_ttend(icol) += qt_tend * pd_icol(k) / g;
```

Comment:
I think we need reductions to compute these columns sums. We can run into trouble if different threads are attempting to update the same value.

You can do an arbitrary amount of reducers in the same call
```
Real sum1, sum2, ...;
Kokkos::parallel_reduce(Kokkos::TeamVectorRange(team, opt_tm_grad_lev, num_levs),
                        [&](int k, Real& local_sum1, Real& local_sum2, ...) {
  local_sum1 += ...;
  .
  .
  .
}, sum1, sum2, ...);
```
And the after well need a Kokkos:single to add in the column sums
```
Kokkos::single(Kokkos::PerTeam(team), [&]() {
  o_tl_caret(icol) += sum1;
  .
  .
  .
});
```

Reply:
I believe it overwrites the value you already have set. You could do the reduction (as you have it written) and then add in the other part, but you would still need a Kokkos::single.

Reply:
If `wp` contains stuff you want to keep, then I think so. I need to check this, maybe if you only hand a scalar `sum` (and not a `Kokkos::Sum<Real>(sum)`) as the last arg it won't overwrite? But I'm 95% sure it used the default `init` for sum reduction which sets the value to 0 at the start.

Also, that team_barrier is unnecessary since it's the end of the kernel.

Reply:
The only issue in your current code is that you set the value 
```
o_tl_caret(icol) = initial_value
```
and then want to add a column sum to that,
```
o_tl_caret(icol) = initial_value + column_val(opt_tm_grad_lev) + ... + column_val(nlevs-1)
``` 
but the `parallel_reduce()` is going to overwrite `o_tl_caret(icol)` to just be whatever the result of the reduction is, so you lose any prior info you had
```
o_tl_caret(icol) = 0 + column_val(opt_tm_grad_lev) + ... + column_val(nlevs-1)
```

In the water path example, if you want 
```
wp(icol) = q_icol(0)*rho_icol(0)/g + ... + q_icol(nlevs-1)*rho_icol(nlevs-1)/g
```
then it is doing exactly what you want.

Reply:
ahh I completely read it wrong!

Reply:
My bad! I was just looking at another output variable. So yes, you will need a reduction, but can just hand in the o_tl_caret(icol) to the reducer.